### PR TITLE
ReadableBuffer cleanup and changes

### PR DESCRIPTION
**Pull request by davidfowl**
- Make first computed to save some room
- Move Equals and StartsWith to extensions

/cc @benaadams I took the computed first from https://github.com/dotnet/corefxlab/pull/1277/files

### CODE REVIEW

File: src/System.IO.Pipelines/ReadableBuffer.cs [hunk 1]
```diff
@@ -56,8 +59,8 @@ internal ReadableBuffer(ReadCursor start, ReadCursor end)
         {
             _start = start;
             _end = end;
-            start.TryGetBuffer(end, out _first);
-            _length = -1;
+            _length = start.GetLength(_end);
```
Comment (benaadams) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: `_length = start.GetLength(end);`? Skip the RAW hazard? Probably doesn't matter but [hazards](https://en.wikipedia.org/wiki/Hazard_(computer_architecture)#Read_after_write_.28RAW.29) keep me up at night

**Comment by benaadams** [comment]
Would be major rework; but once length is calculated you could throw away the cursors and trim to 16 bytes
```csharp
public struct ReadableBuffer 
{
    private BufferSegment _segment;
    private int _index;
    private int _length;
}
``` 
Then traverse the segments with length rather than checking whether `== ._end`

**Comment by benaadams** [comment]
Though it depends how much `End` cursor is used (and whether it can be replaced with length); as reconstructing it would be more expensive.

File: src/System.IO.Pipelines.Extensions/ReadableBufferExtensionsExtensions.cs [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 namespace System.IO.Pipelines
 {
-    public static class TrySliceToExtensions
+    public static class ReadableBufferExtensionsExtensions
```
Comment (pakrym) [inline src/System.IO.Pipelines.Extensions/ReadableBufferExtensionsExtensions.cs]: `ExtensionsExtensions`

File: src/System.IO.Pipelines/ReadableBuffer.cs [hunk 3]
```diff
@@ -40,7 +36,14 @@ public struct ReadableBuffer : ISequence<ReadOnlyMemory<byte>>
         /// <summary>
         /// The first <see cref="Memory{Byte}"/> in the <see cref="ReadableBuffer"/>.
         /// </summary>
-        public Memory<byte> First => _first;
+        public Memory<byte> First
```
Comment (pakrym) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: Make this a method so it would be more obvious that this is not cached?

Comment (davidfowl) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: Don't want I can add a method and deprecate this so it doesn't break immediately.

File: samples/System.IO.Pipelines.Samples/HttpClient/LibuvHttpClientHandler.cs [hunk 4]
```diff
@@ -147,18 +147,22 @@ public LibuvHttpClientHandler()
...
                             // Check for final CRLF.
                             responseBuffer = responseBuffer.Slice(1);
-                            ch = responseBuffer.Peek();
+                            if (responseBuffer.Length == 0)
+                            {
+                                ch = responseBuffer.First.Span[0];
+                            }
                             responseBuffer = responseBuffer.Slice(1);
 
                             if (ch == -1)
```
Comment (benaadams) [inline samples/System.IO.Pipelines.Samples/HttpClient/LibuvHttpClientHandler.cs]: Will be == '\r' if `length > 0` and something else `if length == 0`? But never -1?

Comment (davidfowl) [inline samples/System.IO.Pipelines.Samples/HttpClient/LibuvHttpClientHandler.cs]: Crap, I messed this up. I need to rewrite all of this to use the `ReadableBufferReader`

File: src/System.IO.Pipelines/ReadableBuffer.cs (hunk 3 above)
Comment (pakrym) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: Should've done this. :)

Comment (davidfowl) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: This made all of the code ugly. Filed this https://github.com/dotnet/corefxlab/issues/1285

Comment (davidfowl) [inline src/System.IO.Pipelines/ReadableBuffer.cs]: I'll do it though. I want to rewrite a bunch of the samples.

**Comment by mjp41** [comment]
Does it make sense to expose a FirstSpan? A lot of uses do First.Span...?

**Comment by davidfowl** [comment]
> Does it make sense to expose a FirstSpan? A lot of uses do First.Span...?

That's not bad. The idea here is to deter people from using this type for looking at the data. Just get the span and pass it all the way down the stack. https://github.com/aspnet/KestrelHttpServer/blob/13519b6079043f6b800b70011175912abed57e19/src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs#L34-L67

Get the span once then party on it and pass it down as long as possible. 

For more complex parsing, you make a `ReadableBufferReader` and enumerate spans on that directly:

https://github.com/aspnet/KestrelHttpServer/blob/13519b6079043f6b800b70011175912abed57e19/src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs#L260-L378

**Pull request merged**